tests/distributions/test_transforms.py

```diff
@@ -81,208 +83,97 @@ def _test_shape(self, base_shape, transform):
         sample = dist.TransformedDistribution(base_dist, [transform]).sample()
         assert sample.shape == base_shape
 
-    def test_batchnorm_jacobians(self):
-        for input_dim in [2, 5, 10]:
+    def _test(self, transform_factory, shape=True, jacobian=True, inverse=True):
+        for input_dim in [2, 5, 10]:
+            transform = transform_factory(input_dim)
+            if jacobian:
+                self._test_jacobian(input_dim, transform)
+            if inverse:
+                self._test_inverse(input_dim, transform)
+            if shape:
+                for shape in [(3,), (3, 4)]:
+                    self._test_shape(shape + (input_dim,), transform)
+
+    def _test_conditional(self, conditional_transform_factory, context_dim=3, **kwargs):
+        def transform_factory(input_dim, context_dim=context_dim):
+            z = torch.rand(context_dim)
+            return conditional_transform_factory(input_dim, context_dim).condition(z)
+        self._test(transform_factory, **kwargs)
+
+    # Affine autoregressive
+    def test_affine_autoregressive(self):
+        for stable in [True, False]:
+            self._test(partial(T.affine_autoregressive, stable=stable))
+
+    # Affine Coupling
+    def test_affine_coupling(self):
+        self._test(T.affine_coupling)
+
+    # Batchnorm
+    def test_batchnorm(self):
```

Comment:
what's going on in `test_batchnorm`?

Reply:
From the docs, the reason we need `transform.eval()` is explained by:

> Also, when the module is set to training mode, the sample mean and variance on the current minibatch are used in place of the smoothed averages, :math:`\\hat{\\mu}` and :math:`\\hat{\\sigma^2}`, for the inverse operator. For this reason it is not the case that :math:`x=g(g^{-1}(x))` during training, i.e., that the inverse operation is the inverse of the forward one.

I'm not entirely certain if the following line is necessary: `transform._inverse(torch.normal(torch.arange(0., input_dim), torch.arange(1., 1. + input_dim) / input_dim))`. What it does is populate the smoothed averages

Reply:
I think populating the smoothed averages is necessary actually since if you don't do that then they are all 0's and 1's, and if all dimensions of the smoothed average are the same this could obscure an error in the transform